paasta_tools/autoscaling/autoscaling_cluster_lib.py

```diff
@@ -321,26 +340,92 @@ def gracefully_terminate_slave(self, slave_to_kill, current_capacity, new_capaci
     def filter_aws_slaves(self, slaves_list):
         ips = self.get_instance_ips(self.instances, region=self.resource['region'])
         self.log.debug("IPs in AWS resources: {}".format(ips))
-        slaves = [slave for slave in slaves_list if slave_pid_to_ip(slave['task_counts'].slave['pid']) in ips]
+        slaves = [
+            slave for slave in slaves_list
+            if slave_pid_to_ip(slave['task_counts'].slave['pid']) in ips
+        ]
         slave_ips = [slave_pid_to_ip(slave['task_counts'].slave['pid']) for slave in slaves]
-        instance_descriptions = self.describe_instances([], region=self.resource['region'],
-                                                        instance_filters=[{'Name': 'private-ip-address',
-                                                                           'Values': slave_ips}])
         instance_type_weights = self.get_instance_type_weights()
-        slave_instances = [PaastaAwsSlave(slave, instance_descriptions, instance_type_weights) for slave in slaves]
-        return slave_instances
+        instance_statuses = self.instance_status_for_ips(slave_ips)
+        instance_descriptions = self.instance_descriptions_for_ips(slave_ips)
+
+        paasta_aws_slaves = []
+        for slave in slaves:
```

Comment:
Would this be any better if there was a function called

    get_ec2_description_for_mesos_slave

that was a memoized version of what you have here in lines 354-372

then the code would just need to iterate over each slave, but call the function in line 377 ?


I think that would simplify the code for readability sake, yet still allow it to be fast maybe?